src/compiler/types.ts

```diff
@@ -8035,6 +8035,62 @@ export type UniqueNameHandler = (baseName: string, checkFn?: (name: string) => b
 
 export type EmitHelperUniqueNameCallback = (name: string) => string;
 
+/**
+ * Indicates the minimum `ScriptTarget` (inclusive) after which a specific language feature is no longer transpiled.
+ *
+ * @internal
+ */
+export namespace LanguageFeatureMinimumTarget {

```

Comment:
Can we turn this into a const enum such that it gets inlined? I believe we should be able to reference ScriptTarget from one. Or does that fail because of the type mismatch with ScriptTarget?

(Would like to limit our use of namespaces, honestly... plus potential perf hits here)

Reply:
Or just a big `const` since we don't really need the reverse lookup

Reply:
It doesn't fail to match `ScriptTarget`, but it does produce its own type which doesn't illustrate the relationship when looking at the `.d.ts` file:

![image](https://github.com/microsoft/TypeScript/assets/3902892/cf9829af-6a39-458b-a98e-e18ce739beda)

vs.

![image](https://github.com/microsoft/TypeScript/assets/3902892/92d2c7de-0c62-4439-ac27-04e1ec1cc9c8)

vs.

![image](https://github.com/microsoft/TypeScript/assets/3902892/bcafd9c8-83cd-4fbe-bfc9-7e1d8325a3c3)


I actually still like `namespace` in moderation. This felt like an appropriate use for it, but I don't mind the `as const`. There's no perf difference between `namespace` and `as const`.

Reply:
This is an internal declaration, so I don't personally feel like it's a big deal for the values to be opaque. It does make me wonder if we're missing something in declaration emit that would have reused those nodes, though.

Reply:
Namespaces, as a concept, exist in both the standard library (`Math`, `Atomics`, `Reflect`) and the web platform (`subtle`, `performance`, etc.) so I don't agree that they should be wholly discouraged.

Reply:
In type declarations, sure, I'm just trying to avoid the runtime cost of them (e.g. const enums are inlined, modules are tree shaken).

Reply:
If the perf results show a significant dip, I can either make it a const enum, it just feels less useful when hover information only shows the final value like `2` instead of something meaningful like `ES2015`.

Reply:
I see what you mean; I do wish our hover and declaration emit did better here. But, we seem to just always compute in dts emit (not sure about hover).

Reply:
> I see what you mean; I do wish our hover and declaration emit did better here. But, we seem to just always compute in dts emit (not sure about hover).

It's arguable as to whether someone who writes `const enum E2 { A = E1.B }` would expect hover/d.ts to print the value or qualified name. With `const` variables or fields it's far clearer as to what to show. I could see maybe supporting `const enum E2 { A = E1.B as E1.B }` or something similar to explicitly opt-in to showing the qualified name, but that seems like a very niche use case.

Reply:
I would much rather see `(enum member) TypeFlags.Unit = Enum | Literal | UniqueESSymbol | Nullable` than `(enum member) TypeFlags.Unit = 109472`. The actual computed values seem like the least important part of an enum.